safe/impact_functions/inundation/flood_raster_road/impact_function.py

```diff
@@ -0,0 +1,394 @@
+    ct = QgsCoordinateTransform(raster.crs(), roads_crs)
+
+    for y in xrange(raster_rows):
+        for x in xrange(raster_cols):
+            # only use cells that are within the specified threshold
+            value = block.value(y, x)
+            if value < threshold_min or value > threshold_max:
+                continue
+
+            # construct rectangular polygon feature for the cell
+            x0 = raster_xmin+(x*cell_width)
+            x1 = raster_xmin+((x+1)*cell_width)
+            y0 = raster_ymax-(y*cell_height)
+            y1 = raster_ymax-((y+1)*cell_height)
+            outer_ring = [QgsPoint(x0, y0), QgsPoint(x1, y0),
+                          QgsPoint(x1, y1), QgsPoint(x0, y1),
+                          QgsPoint(x0, y0)]
+            geometry = QgsGeometry.fromPolygon([outer_ring])
+            geometry.transform(ct)
+            f = QgsFeature()
+            f.setGeometry(geometry)
+            features.append(f)
+
+    _, features = vl.dataProvider().addFeatures(features)
+
+    # construct a temporary map for fast access to features by their IDs
+    # (we will be getting feature IDs from spatial index)
+    flood_cells_map = {}
+    for f in features:
+        flood_cells_map[f.id()] = f
+
+    # build a spatial index so we can quickly identify
+    # flood cells overlapping roads
+    if QGis.QGIS_VERSION_INT >= 20800:
+        # woohoo we can use bulk insert (much faster)
+        index = QgsSpatialIndex(vl.getFeatures())
+    else:
+        index = QgsSpatialIndex()
+        for f in vl.getFeatures():
+            index.insertFeature(f)
+
+    return index, flood_cells_map
+
+
+def _add_output_feature(features, geom, is_flooded,
+                        fields, orig_attributes, target_field):
+    """ Utility function to construct road features from geometry.
+
+    Newly created features get the attributes from the original feature.
+    If the geometry is multi-part, it will be exploded into several
+    single-part features.
+    """
+    geoms = geom.asGeometryCollection() if geom.isMultipart() else [geom]
+    for g in geoms:
+        f = QgsFeature(fields)
+        f.setGeometry(g)
+        for attr_no, attr_val in enumerate(orig_attributes):
+            f.setAttribute(attr_no, attr_val)
+        f.setAttribute(target_field, is_flooded)
+        features.append(f)
+
+
+def _union_geometries(geoms):
+    """ Return a geometry which is union of the passed list of geometries """
+    if QGis.QGIS_VERSION_INT >= 20400:
+        # woohoo we can use fast union (needs GEOS >= 3.3)
+        return QgsGeometry.unaryUnion(geoms)
+    else:
+        # uhh we need to use slow iterative union
+        if len(geoms) == 0:
+            return QgsGeometry()
+        result_geometry = QgsGeometry(geoms[0])
+        for g in geoms[1:]:
+            result_geometry = result_geometry.combine(g)
+        return result_geometry
+
+
+def _intersect_roads_flood(roads, request, index,
```

Comment:
Can you make a unit test for this please? I also think we should move this to one of our utility classes e.g. safe/gis/qgis_vector_tools.py